### PR TITLE
CockroachDB v21 Support

### PR DESCRIPTION
**Pull request by keithdoggett**
I'm going to track any changes that come up in the v21 betas and RCs here. If any of the changes don't apply to the 6.0.x or 5.2.x branches, I'll open up separate PRs for them.

#### Last Tested Version

v21.1.0

#### Changes

- Change create_unlogged_tables_test to reflect the `relpersistence` field for temporary tables in `pg_class` catalog being changed from "p" to "t".
- Add support for comments.

### CODE REVIEW

**Comment by keithdoggett** [comment]
The only test failure was one of the random fixture foreign key issues that pop up sometimes:

<details><summary>Log</summary>
<p>

```
[15:45:49]
[Step 2/2] Error:
[15:45:49]
[Step 2/2] IgnoreFixturesTest#test_ignores_books_fixtures:
[15:45:49]
[Step 2/2] ActiveRecord::InvalidForeignKey: PG::ForeignKeyViolation: ERROR:  foreign key violation: "parrots_treasures" row treasure_id=25772709, rowid=659562000003104769 has no match in "treasures"
[15:45:49]
[Step 2/2] 
[15:45:49]
[Step 2/2]     /home/agent/.gems/bundler/gems/rails-6ce0ce921d5b/activerecord/lib/active_record/connection_adapters/postgresql/database_statements.rb:47:in `exec'
[15:45:49]
[Step 2/2]     /home/agent/.gems/bundler/gems/rails-6ce0ce921d5b/activerecord/lib/active_record/connection_adapters/postgresql/database_statements.rb:47:in `block (2 levels) in execute'
[15:45:49]
[Step 2/2]     /home/agent/.gems/bundler/gems/rails-6ce0ce921d5b/activesupport/lib/active_support/dependencies/interlock.rb:48:in `block in permit_concurrent_loads'
[15:45:49]
[Step 2/2]     /home/agent/.gems/bundler/gems/rails-6ce0ce921d5b/activesupport/lib/active_support/concurrency/share_lock.rb:187:in `yield_shares'
[15:45:49]
[Step 2/2]     /home/agent/.gems/bundler/gems/rails-6ce0ce921d5b/activesupport/lib/active_support/dependencies/interlock.rb:47:in `permit_concurrent_loads'
[15:45:49]
[Step 2/2]     /home/agent/.gems/bundler/gems/rails-6ce0ce921d5b/activerecord/lib/active_record/connection_adapters/postgresql/database_statements.rb:46:in `block in execute'
[15:45:49]
[Step 2/2]     /home/agent/.gems/bundler/gems/rails-6ce0ce921d5b/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:696:in `block (2 levels) in log'
[15:45:49]
[Step 2/2]     /home/agent/.gems/bundler/gems/rails-6ce0ce921d5b/activesupport/lib/active_support/concurrency/load_interlock_aware_monitor.rb:26:in `block (2 levels) in synchronize'
[15:45:49]
[Step 2/2]     /home/agent/.gems/bundler/gems/rails-6ce0ce921d5b/activesupport/lib/active_support/concurrency/load_interlock_aware_monitor.rb:25:in `handle_interrupt'
[15:45:49]
[Step 2/2]     /home/agent/.gems/bundler/gems/rails-6ce0ce921d5b/activesupport/lib/active_support/concurrency/load_interlock_aware_monitor.rb:25:in `block in synchronize'
[15:45:49]
[Step 2/2]     /home/agent/.gems/bundler/gems/rails-6ce0ce921d5b/activesupport/lib/active_support/concurrency/load_interlock_aware_monitor.rb:21:in `handle_interrupt'
[15:45:49]
[Step 2/2]     /home/agent/.gems/bundler/gems/rails-6ce0ce921d5b/activesupport/lib/active_support/concurrency/load_interlock_aware_monitor.rb:21:in `synchronize'
[15:45:49]
[Step 2/2]     /home/agent/.gems/bundler/gems/rails-6ce0ce921d5b/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:695:in `block in log'
[15:45:49]
[Step 2/2]     /home/agent/.gems/bundler/gems/rails-6ce0ce921d5b/activesupport/lib/active_support/notifications/instrumenter.rb:24:in `instrument'
[15:45:49]
[Step 2/2]     /home/agent/.gems/bundler/gems/rails-6ce0ce921d5b/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:687:in `log'
[15:45:49]
[Step 2/2]     /home/agent/.gems/bundler/gems/rails-6ce0ce921d5b/activerecord/lib/active_record/connection_adapters/postgresql/database_statements.rb:45:in `execute'
[15:45:49]
[Step 2/2]     /home/agent/.gems/bundler/gems/rails-6ce0ce921d5b/activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb:1066:in `add_foreign_key'
[15:45:49]
[Step 2/2]     /home/agent/activerecord-cockroachdb-adapter/lib/active_record/connection_adapters/cockroachdb/referential_integrity.rb:25:in `block in disable_referential_integrity'
[15:45:49]
[Step 2/2]     /home/agent/activerecord-cockroachdb-adapter/lib/active_record/connection_adapters/cockroachdb/referential_integrity.rb:23:in `each'
[15:45:49]
[Step 2/2]     /home/agent/activerecord-cockroachdb-adapter/lib/active_record/connection_adapters/cockroachdb/referential_integrity.rb:23:in `disable_referential_integrity'
[15:45:49]
[Step 2/2]     /home/agent/activerecord-cockroachdb-adapter/lib/active_record/connection_adapters/cockroachdb/database_statements.rb:26:in `block in insert_fixtures_set'
[15:45:49]
[Step 2/2]     /home/agent/activerecord-cockroachdb-adapter/lib/active_record/connection_adapters/cockroachdb/database_statements.rb:97:in `with_multi_statements'
[15:45:49]
[Step 2/2]     /home/agent/activerecord-cockroachdb-adapter/lib/active_record/connection_adapters/cockroachdb/database_statements.rb:25:in `insert_fixtures_set'
[15:45:49]
[Step 2/2]     /home/agent/.gems/bundler/gems/rails-6ce0ce921d5b/activerecord/lib/active_record/fixtures.rb:638:in `block in insert'
[15:45:49]
[Step 2/2]     /home/agent/.gems/bundler/gems/rails-6ce0ce921d5b/activerecord/lib/active_record/fixtures.rb:629:in `each'
[15:45:49]
[Step 2/2]     /home/agent/.gems/bundler/gems/rails-6ce0ce921d5b/activerecord/lib/active_record/fixtures.rb:629:in `insert'
[15:45:49]
[Step 2/2]     /home/agent/.gems/bundler/gems/rails-6ce0ce921d5b/activerecord/lib/active_record/fixtures.rb:615:in `read_and_insert'
[15:45:49]
[Step 2/2]     /home/agent/.gems/bundler/gems/rails-6ce0ce921d5b/activerecord/lib/active_record/fixtures.rb:567:in `create_fixtures'
[15:45:49]
[Step 2/2]     /home/agent/.gems/bundler/gems/rails-6ce0ce921d5b/activerecord/lib/active_record/test_fixtures.rb:227:in `load_fixtures'
[15:45:49]
[Step 2/2]     /home/agent/.gems/bundler/gems/rails-6ce0ce921d5b/activerecord/lib/active_record/test_fixtures.rb:156:in `setup_fixtures'
[15:45:49]
[Step 2/2]     /home/agent/.gems/bundler/gems/rails-6ce0ce921d5b/activerecord/lib/active_record/test_fixtures.rb:10:in `before_setup'
[15:45:49]
[Step 2/2] 
[15:45:49]
[Step 2/2] rails test home/agent/.gems/bundler/gems/rails-6ce0ce921d5b/activerecord/test/cases/fixtures_test.rb:1311
```

</p>
</details>

If this looks good I can try to run the tests again and see if we can get it to pass.

File: test/excludes/CommentTest.rb
Comment (rafiss) [inline test/excludes/CommentTest.rb:1]: ALTER COLUMN TYPE is supported with the setting `SET enable_experimental_alter_column_type_general = true`. we could turn that setting on for these tests.

Comment (keithdoggett) [reply test/excludes/CommentTest.rb:1]: I tried to run the tests with this set and they ended up failing for another reason. In both tests, they change a column type from `bool` to `character varying`, and once a column is altered, the comment on it can't be changed.

I ended up modifying the tests to not change the column type since these are mainly testing that comments can be removed/changed through migrations. 

I can open an issue about the inability to modify comments on altered columns if it's not already tracked.

Comment (rafiss) [reply test/excludes/CommentTest.rb]: oh nice find! please do open that issue, that sounds like an unknown bug. if you can post SQL to reproduce it that would help.

**Review by rafiss — APPROVED** [review]
lgtm!

> random fixture foreign key issues that pop up sometimes

at some point (maybe once some of the in-flight issues die down) i'd love if we could figure out the cause of these random failures. (or have you done that already?)

File: test/excludes/CommentTest.rb
Comment (keithdoggett) [reply test/excludes/CommentTest.rb]: https://github.com/cockroachdb/cockroach/issues/65568

**Comment by keithdoggett** [comment]
Unfortunately, no I have not solved that issue yet, but I agree once we get through some of these outstanding issues it would be a good idea to look into that.

**Comment by vy-ton** [comment]
@keithdoggett What's the status of getting CockroachDB 21.1 support merged?

**Comment by keithdoggett** [comment]
@vy-ton this PR was approved, so I think we should be good to merge this in and backport it, unless there's some other blocker I'm not aware of. @rafiss do you know if there's anything else we need to fix to get this merged?

**Review by rafiss — CHANGES_REQUESTED** [review]
i think it's ready, but just noticed one small thing in the tests!

**Comment by keithdoggett** [comment]
Sorry, what do you want to be fixed in the tests? I'm not seeing if you marked it in the changes.

File: test/cases/create_unlogged_tables_test.rb
Comment (rafiss) [inline test/cases/create_unlogged_tables_test.rb:56]: it seems like we can just remove this `test_gracefully_handles_temporary_tables` method entirely, and go back to using the upstream test.

I think we have to keep `test_unlogged_in_test_environment_when_unlogged_setting_enabled`, but can you adjust the comment on that one to point at https://github.com/cockroachdb/cockroach/pull/56827 and mention that CockroachDB ignores the `UNLOGGED` specifier

**Review by rafiss — CHANGES_REQUESTED** [review]

**Comment by keithdoggett** [comment]
Looks like this is getting hung on something. I'll try running the full suite locally to see if I can reproduce it. Is there anyway to restart the tests in the CI or does it just have to wait for another push?

File: test/cases/create_unlogged_tables_test.rb
Comment (keithdoggett) [reply test/cases/create_unlogged_tables_test.rb]: This is actually failing once removed since we need the connection to set `experimental_enable_temp_tables`. I could potentially set that earlier in the testing flow (or in the adapter), but it would probably be easier to leave it and track this setting somewhere.

Comment (rafiss) [reply test/cases/create_unlogged_tables_test.rb]: Ah ok, sorry for the distraction then! I thought we had done something to enable temp tables everywhere. (It can be done globally by using `SET CLUSTER SETTING sql.defaults.experimental_temp_tables_enabled = true`)

Comment (keithdoggett) [reply test/cases/create_unlogged_tables_test.rb]: Hmm, I tried setting that in the teamcity-test build with 

```
cockroach sql --insecure -e "SET CLUSTER SETTING sql.defaults.experimental_temp_tables_enabled = true"
```

but it raised an error

```
ERROR: unknown cluster setting 'sql.defaults.experimental_temp_tables_enabled
```

I also couldn't find it in the result from `SHOW CLUSTER SETTINGS`.

Comment (keithdoggett) [reply test/cases/create_unlogged_tables_test.rb:55]: I just reverted it for now, but once we figure out the setting I can remove that test.

Comment (rafiss) [reply test/cases/create_unlogged_tables_test.rb:55]: my bad.. i shouldn't have tried saying the name from memory.

it's `SET CLUSTER SETTING sql.defaults.experimental_temporary_tables.enabled = 'true';`

the next release of cockroachdb will include it in `SHOW CLUSTER SETTINGS`

**Review by rafiss — APPROVED** [review]
this all lgtm now! i'll let you decide if you want to do the temp tables setting now or later, but feel free to merge when you like

File: test/cases/create_unlogged_tables_test.rb
Comment (keithdoggett) [reply test/cases/create_unlogged_tables_test.rb]: Ok cool that worked thanks! I added that to the teamcity-test setup and removed the `test_gracefully_handles_temporary_tables` test.

**Comment by keithdoggett** [comment]
It looks like the jobs are still hanging even with the 3-minute time limit on each test. Should we try to figure that out before merging?

```
[03:24:15]
[Step 2/2] ActiveRecord::PostgresqlTransactionNestedTest#test_unserializable_transaction_raises_SerializationFailure_inside_nested_SavepointTransaction = 180.10 s = F
[03:24:15]
[Step 2/2] 
[03:24:15]
[Step 2/2] 
[03:24:15]
[Step 2/2] Failure:
[03:24:15]
[Step 2/2] ActiveRecord::PostgresqlTransactionNestedTest#test_unserializable_transaction_raises_SerializationFailure_inside_nested_SavepointTransaction [/home/agent/.gems/bundler/gems/rails-f94f41dd59b5/activerecord/test/cases/adapters/postgresql/transaction_nested_test.rb:39]:
[03:24:15]
[Step 2/2] [ActiveRecord::SerializationFailure] exception expected, not
[03:24:15]
[Step 2/2] Class: <Timeout::Error>
[03:24:15]
[Step 2/2] Message: <"Test took over 3 minutes to finish">
[03:24:15]
[Step 2/2] ---Backtrace---
[03:24:15]
[Step 2/2] /home/agent/.gems/bundler/gems/rails-f94f41dd59b5/activerecord/test/cases/adapters/postgresql/transaction_nested_test.rb:66:in `join'
[03:24:15]
[Step 2/2] /home/agent/.gems/bundler/gems/rails-f94f41dd59b5/activerecord/test/cases/adapters/postgresql/transaction_nested_test.rb:66:in `block (2 levels) in <class:PostgresqlTransactionNestedTest>'
```

per the logs, it seems like the timeout is working, but the test suite isn't finishing. Could it be something going on with the CI?

**Comment by rafiss** [comment]
Postgres implements SERIALIZABLE differently than CockroachDB, so I'm not surprised that test is timing out. Skipping makes sense.

But I don't have an idea for why the test timing out would cause the whole CI to hang. To me, it doesn't feel worth debugging, since we want to skip the test anyway. So this looks good to merge!

**Comment by Piioo** [comment]
When you will create a new version with this changes?

**Pull request merged**
